### PR TITLE
BFD-2359: Decouple Deployment from Build Delivery Deployment Pipleine

### PR DESCRIPTION
**Pull request by JeremyIR**
**JIRA Ticket:**
[BFD-2359](https://jira.cms.gov/browse/BFD-2359)

**User Story or Bug Summary:**

As a BFD engineer, I want the ability to run an isolated job for deploying the ultimate BFD artifacts to targeted environments so that I can avoid having hasty workarounds or interference to other engineers working in the bfd-test project.
---

### What Does This PR Do?

Deploying bfd is something that should be isolated, so that other engineers can get their work done more harmoniously in an organized, efficient fashion. This PR introduces such a pipeline that allows for a git branch/reference, ami IDs, and environment as an input. This is eventually going to lead to deploying to ephemeral environments so other engineers wont have to worry about who is using test, prod-sbx, etc.

### What Should Reviewers Watch For?

Is the app/ami deployment with a custom branch logic in the new pipeline sound? 

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [x] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [x] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [x] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [x] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] The data dictionary has been updated with any field mapping changes, if any were made.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: ops/jenkins/bfd-deploy-apps/Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+                script {
+                    currentStage = env.STAGE_NAME
+                    try {
+                        def mgiratorAmiOverride = (params.migrator_ami_override?.trim())
+                                                        ? params.migrator_ami_override.trim()
+                                                        : ""
+
+                        lock(resource: lockResource) {
+                            migratorDeploymentSuccessful = migratorScripts.deployMigrator(
+                                amiId: mgiratorAmiOverride,
```
Comment (mjburling) [inline ops/jenkins/bfd-deploy-apps/Jenkinsfile:206]:
```suggestion
                        def migratorAmiOverride = (params.migrator_ami_override?.trim())
                                                        ? params.migrator_ami_override.trim()
                                                        : ""

                        lock(resource: lockResource) {
                            migratorDeploymentSuccessful = migratorScripts.deployMigrator(
                                amiId: migratorAmiOverride,
```

Aside from the typo... I'm a little surprised by this. I haven't given this enough thought, but maybe you can help–how does this behave when `deployMigrator()` receives an empty string for the ami override?

Comment (JeremyIR) [reply ops/jenkins/bfd-deploy-apps/Jenkinsfile]: Decided to revert this as it seems a bit excessive.

**Review by malessi — CHANGES_REQUESTED** [review]
Couple of minor things that I noticed, but overall looks good! I've left some suggestions for changes that should make this pipeline a bit more robust, and added some minor nits for more stylistic things.

File: ops/jenkins/bfd-deploy-apps/Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+                    currentStage = env.STAGE_NAME
+                    try {
+                        def pipelineAmiOverride = [ami_id_override: params.pipeline_ami_override.trim()]
+                        lock(resource: lockResource) {
+                            awsAuth.assumeRole()
+                            terraform.deployTerraservice(
+                            env: trimmedEnv,
+                            directory: 'ops/terraform/services/pipeline',
+                            tfVars: pipelineAmiOverride
+                            )
```
[the remark this answers is not in the document]
Comment (JeremyIR) [reply ops/jenkins/bfd-deploy-apps/Jenkinsfile]: great suggestion.

File: ops/jenkins/bfd-deploy-apps/Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+        )
+
+        string(
+        name: 'pipeline_ami_override',
+        description: 'The AMI ID to deploy pipeline from',
+        defaultValue: null
+        )
+
+        string(
+        name: 'server_ami',
```
Comment (billie-rose) [inline ops/jenkins/bfd-deploy-apps/Jenkinsfile:117]:
```suggestion
        name: 'server_ami_override',
```

File: ops/jenkins/bfd-deploy-apps/Jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+        description: 'The AMI ID to deploy server from',
+        defaultValue: null
+        )
+
+        booleanParam(
+            name: 'force_migrator_deployment', 
+            description: 'When true, force the migrator to deploy.',
```
Comment (billie-rose) [inline ops/jenkins/bfd-deploy-apps/Jenkinsfile:124]:
```suggestion
            description: 'When true, force the migrator to deploy', 
```

(no period, or add periods to the other descriptions) 😉

File: ops/jenkins/bfd-deploy-apps/Jenkinsfile [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+                            } else {
+                                println "See ${migratorRunbookUrl} for troubleshooting resources."
+                                error('Migrator deployment failed')
+                            }
+                        }
+                      } catch (org.jenkinsci.plugins.workflow.steps.FlowInterruptedException e) {
+                        currentBuild.result = 'ABORTED'
+                        sendNotifications(trimmedEnv, currentBuild.currentResult, currentStage, gitCommitId, gitRepoUrl)
+                        throw e
+                    } catch (ex) {
```
Comment (billie-rose) [inline ops/jenkins/bfd-deploy-apps/Jenkinsfile:225]: formatting seems off here - let's format the whole file if possible

File: ops/jenkins/bfd-deploy-apps/Jenkinsfile [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+                }
+            }
+        }
+
+        stage('Deploy Server') {
+            steps {
+                script {
+                    currentStage = env.STAGE_NAME
+                    try {
+                        def serverAmi = params.server_ami.trim()
```
Comment (billie-rose) [inline ops/jenkins/bfd-deploy-apps/Jenkinsfile:268]:
```suggestion
                        def serverAmi = params.server_ami?.trim()
```

File: ops/jenkins/bfd-deploy-apps/Jenkinsfile (hunk 3 above)
Comment (JeremyIR) [reply ops/jenkins/bfd-deploy-apps/Jenkinsfile:117]: BFD Server isn't yet an explicit Terraservice and is instead deployed as part of stateless, so this really isn't an override and the operator/runner is required to specify the BFD Server AMI.

**Comment by billie-rose** [comment]
minor comments - nice work!

File: ops/jenkins/bfd-deploy-apps/Jenkinsfile [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+    parameters {
+        string(
+        name: 'env',
+        description: 'The BFD Environment to deploy apps to'
+        )
+
+        string(
+        name: 'migrator_ami_override',
+        description: 'The AMI ID to deploy migrator from',
+        defaultValue: null
```
[the remark this answers is not in the document]
Comment (JeremyIR) [reply ops/jenkins/bfd-deploy-apps/Jenkinsfile:107]: No default value needed

File: ops/jenkins/bfd-deploy-apps/Jenkinsfile (hunk 5 above)
Comment (JeremyIR) [reply ops/jenkins/bfd-deploy-apps/Jenkinsfile:225]: fixed

File: ops/jenkins/bfd-deploy-apps/Jenkinsfile (hunk 7 above)
[the remark this answers is not in the document]
Comment (billie-rose) [reply ops/jenkins/bfd-deploy-apps/Jenkinsfile:107]: Can we omit the assignments to `null` then? (I genuinely don't know)

[the remark this answers is not in the document]
Comment (JeremyIR) [reply ops/jenkins/bfd-deploy-apps/Jenkinsfile:107]: See Mitch's comment. He addressed this as well

**Review by billie-rose — APPROVED** [review]

**Review by malessi — APPROVED** [review]

**Pull request merged**